New halt implementation

Ducasse opened the pull request:
a new implementation of haltOnce with nice tests showing that 
executing two times in a row the same haltOnce call will bark only the first time and 
that we can have multiple call sites in the same method with each one their own "counter"

Ducasse committed:
feature: reimplement Halt once and once without the need to rely on the compiler, parse, semantic analysis and the rest

Ducasse committed:
tests: now we have tests that really test call sites.

Ducasse commented:
As a bonus no need for the compiler, not reparse and semantical analysis.

Ducasse committed:
no semantic change: recategorisation of test protocols

Ducasse commented:
Working on reimplementing HaltCount: too. 
No need to scan the full memory for senders and in addition reparse the methods :)

guillep commented on src/Debugging-Utils/Halt.class.st:
```diff
@@ -184,24 +193,36 @@ Halt class >> onCountWithBehavior: behaviorBlock [
 	"Stop once and only once the execution of the method containing the expression Halt once. If you need to get it stops another time after the first stop, use Halt resetOnce."
 
-	| callingContext senderContext node |
-	<debuggerCompleteToSender>
```
Maybe the pragma should be kept, because this is used to hint the debugger that this method should be filtered from the shown stack by default (and usually we want to stop at the caller, not here right?

Or maybe it's not needed? What does the debugger show when this halts without the pragma?

guillep replied:
I mean, if you do `Halt once` directly in a method

Ducasse replied:
Tx I thought that I kept it so I will added back

Ducasse committed:
adding annotation to once

Ducasse committed:
Clean: protocol sorting + initialize split + remove dead code

Ducasse committed:
All the tests pass except containsHalt. 
Now there is a bug because we should not be forced to do Halt initializeOnCount before. 
I just commit to get a stable point.

Ducasse committed:
New version. Now indeed the tests requires Halt initializeOnCount. "This is mandatory else running twice the test will fail."

Ducasse committed:
Fixing containsHalt (it is based on the method protocol :( what a brittle idea!))

Ducasse committed:
Fixing testHaltFromCount

Ducasse committed:
Merge branch 'Pharo13' into newHaltImplementation

Ducasse committed:
Split one test into several ones

Ducasse committed:
Merge fa3d519a4eea8ca278e8b5e916518b0687e56990

Ducasse committed:
Improving tests

Ducasse committed:
Improving tests

Ducasse committed:
Rename internal method

Ducasse committed:
Adding haltUpTo: and tests

Ducasse committed:
Revising implementation and adding comments in class

Ducasse committed:
FIxing a test using an old version of the API

Ducasse committed:
Removing haltCountUpTo:

jecisc committed:
Remove Toto
